MQ2GrimGUI.cpp

```diff
@@ -1,505 +1,1510 @@
+		if (ImGui::Begin("GrimGUI##MainWindow", &s_WinSettings.showMainWindow))
+		{
+			int sizeX = static_cast<int>(ImGui::GetWindowWidth());
+			int col = 1;
+			col = sizeX / 100;
+			if (col < 1)
+				col = 1;
+
+			if (ImGui::BeginTable("Window List", col))
+			{
+				ImGui::TableNextRow();
+				ImGui::TableNextColumn();
+				for (const auto& option : options)
+				{
+					if (ImGui::Checkbox(option.label, option.setting))
+						SaveSetting(option.setting, &s_SettingsFile[0]);
+					
+					ImGui::TableNextColumn();
+				}
+				ImGui::EndTable();
+			}
+
+			ImGui::Separator();
 
+			if (ImGui::Button("Config"))
+				s_WinSettings.showConfigWindow = true;
 
+		}
+		ImGui::End();
+	}
 }
 
+#pragma endregion
+
+#pragma region Plugin API
 
-// Called periodically by MQ2
 PLUGIN_API void OnPulse()
 {
 	auto now = std::chrono::steady_clock::now();
 	if (GetGameState() == GAMESTATE_INGAME)
 	{
+		s_IsCaster = CheckCaster();
+		s_LockAllWin = s_WinSettings.lockAllWin ? ImGuiWindowFlags_NoMove : ImGuiWindowFlags_None;
+		s_WindowFlags = s_WinSettings.showTitleBars ? ImGuiWindowFlags_None : ImGuiWindowFlags_NoTitleBar;
+
+		if (s_ShowOutOfGame)
+			s_ShowOutOfGame = false; // reset incase we logged back in. =)
+
+		if (!s_CharIniLoaded)
+			UpdateSettingFile();
+
 		if (now - g_LastUpdateTime >= g_UpdateInterval)
 		{
-			g_CharData.Update();  // Refresh character data at the specified interval
-			g_TargetData.Update();
+			if (pAggroInfo)
+			{
+				s_NumSettings.myAggroPct = pAggroInfo->aggroData[AD_Player].AggroPct;
+				s_NumSettings.secondAggroPct = pAggroInfo->aggroData[AD_Secondary].AggroPct;
+				
+				if (pAggroInfo->AggroSecondaryID)
+					s_SecondAggroName = GetSpawnByID(pAggroInfo->AggroSecondaryID)->DisplayedName;
+				else
+					s_SecondAggroName = "Unknown";
+			}
+
+			GetHeading();
+
 			g_LastUpdateTime = now;
+
+			s_DanNetEnabled = mq::IsPluginLoaded("MQ2DanNet");
 		}
-		if (now - g_LastFlashTime >= g_UpdateFlashInterval)
+
+		// update buff flash timers
+		if (s_NumSettings.flashBuffInterval > 0)
+		{
+			if (now - g_LastBuffFlashTime >= std::chrono::milliseconds(500 - (s_NumSettings.flashBuffInterval * 10)))
+			{
+				s_WinSettings.flashTintFlag = !s_WinSettings.flashTintFlag;
+				g_LastBuffFlashTime = now;
+			}
+		}
+		else
+		{
+			s_WinSettings.flashTintFlag = false;
+		}
+
+		// update combat flash timers
+		if (s_NumSettings.combatFlashInterval > 0)
+		{
+			if (now - g_LastFlashTime >= std::chrono::milliseconds(500 - (s_NumSettings.combatFlashInterval * 10)))
+			{
+				s_WinSettings.flashCombatFlag = !s_WinSettings.flashCombatFlag;
+				g_LastFlashTime = now;
+			}
+		}
+		else
+		{
+			s_WinSettings.flashCombatFlag = false;
+		}
+
+		if (pSpellPicker->SelectedSpell)
 		{
-			b_flashCombat = !b_flashCombat;
-			g_LastFlashTime = now;
+			s_MemSpellName = pSpellPicker->SelectedSpell->Name;
+			MemSpell(pLocalPlayer, fmt::format("{} \"{}\"", s_MemGemIndex, s_MemSpellName).c_str());
```

Comment:
This should be `DoCommandf("/memspell %s \"%s\"", s_MemGemIndex, s_MemSpellName)`

MemSpell is a command function. You shouldn't even be able to call it from here ...